Base/Python/slicer/parameterNodeWrapper/__init__.py

```diff
@@ -17,3 +22,24 @@
     unannotatedType,
 )
 from .wrapper import *
+
+# Register parameterNodeWrapper plugins
+_start = time.time()
+_plugin_suffix = "ParameterNodeWrapperPlugins"
+_discovered_plugins = {
+    name: importlib.import_module(name)
+    for finder, name, ispkg
+    in pkgutil.iter_modules()
+    if name.endswith(_plugin_suffix)
+}
+_elapsed = time.time() - _start
+
+logging.info(f"Discovered {len(_discovered_plugins)} ParameterNodeWrapper plugins [{_elapsed:.2f}s]")
```

Comment:
Since the SubjectHiearchyPlugins is explicitly registered by its associated scripted C++ loader, the current approach worked for both core and extension.

For ParameterNodeWrapperPlugins, we have to perform the import from somewhere, I also considered adding a call like `execute("import AwesomeParameterNodeWrapperPlugins")`in the associated `qSlicerAwesomeModule::setup()` function but that would always require the module to be loaded ....

Alternatively, we could also import the plugin like we do for the vtk, mrml, logic, ... C++ wrapper (where the `slicer` and `vtk`namespace are updated to have the corresponding classes)

Reply:
Re: package name convention

I think expecting a name is consistent with how we also discover and load modules (`qSlicerAwesomeModule...`)

Reply:
> debug level + all at once

Good idea, I will change this.

Reply:
> I also considered adding a call like `pythonManager->executeString("import AwesomeParameterNodeWrapperPlugins")`in the associated `qSlicerAwesomeModule::setup()` function but that would always require the module to be loaded

I think this would be a better solution. Plugins may be required to be registered when the all the dependent modules are already loaded, after the current module is already loaded, but before modules that depend on the current module are loaded. Maybe not this particular plugin type right now, but if we want to use the same plugin discovery and loading mechanism for all type of scripted plugins.

My concerns with the design proposed in this PR:
- We would pollute the scripted module folder with a ...Lib or ...Plugin folder for each type of plugin. We already have so many plugin types (scripted file dialog, file reader, writer, displayable manager, DICOM, Subject Hierearchy plugin, Segment Editor plugin, Segment Statistics plugin, Sample Data, Self Tests, etc.) that having separate folder for each type for each module would be just too much.
- Iterating through all available packages may be also time-consuming. I see that you added logging of how much time it takes and I guess it is very quick, otherwise you would not have chosen this design, but it feels that it could take significant amount of time when there are many packages.
- It may be somewhat fragile because new Python directories may be added later, which would not be re-scanned.

> I think expecting a name is consistent with how we also discover and load modules (qSlicerAwesomeModule...)

Yes, we use special names for scripted file dialog, file reader/writer plugins (and we could use the same for parameter node wrapper plugins), but it is quite brittle (relies on having a specific class name) and hard to discover. I guess there are other places where we rely on specific file or class name patters, but they suffer from the same issues. So, an explicit registration seems like a nicer solution to me.

---

Overall, to avoid proliferation of plugin mechanisms in Slicer (https://xkcd.com/927/), we should either use any existing mechanisms or add one more new mechanism for the very last time that we intend to eventually replace all other mechanisms.

Reply:
> so many top level


Agreed. We have been conveniently adding everything to the top level without any namespacing and it is time to address our design. 


One or our approach for subject hierarchy plugin (and few others) is to have the plugin implemented directly in the scripted module. For example, see ExportAs[^1] scripted module in SlicerMorph. 


This force to have everything in one file which doesn't really scale, to address this for scriped module we have been using `<ModuleName-ish>Lib` convention.


For example, see `HeartValveLib`[^2]


What we are missing is a convention for Python modules provided by C++ module (that applies to module available in both the core and extension).


In the case of the the subject hiearchy plugins, we could have:

- `SubjectHierachyLib` with:

   - `__init__.py`

   - `AbstractScriptedSubjectHierachy.py`

   - `parameterNodeWrapper/__init__.py` 

  - `parameterNodeWrapper/guiConnector.py`


Then the setup function of the loadable module would be responsible for importing the `SubjectHierarchyLib` package



There is also the package namespace approach we could look into ...


https://packaging.python.org/en/latest/guides/packaging-namespace-packages/


[^1]: https://github.com/SlicerMorph/SlicerMorph/blob/master/ExportAs%2FExportAs.py


[^2]: https://github.com/SlicerHeart/SlicerHeart/blob/092fefd9913fb4f394202fe8c8fecde96c2523ca/ValveAnnulusAnalysis/HeartValveLib/HeartValvesSubjectHierarchyPlugin.py#L3